marker_api/src/ast/common/id.rs

```diff
@@ -144,3 +146,63 @@ new_id! {
     /// This ID uniquely identifies a statement during linting.
     pub StmtId: u64
 }
+
+#[repr(C)]
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy)]
+pub enum NodeId {
+    Expr(ExprId),
+    Item(ItemId),
+    Stmt(StmtId),
+    Body(BodyId),
+    Field(FieldId),
+    Variant(VariantId),
+}
+
+macro_rules! impl_into_node_id_for {
+    ($variant:ident, $ty:ty) => {
+        impl From<$ty> for NodeId {
+            fn from(value: $ty) -> Self {
+                NodeId::$variant(value)
+            }
+        }
+
+        impl From<&$ty> for NodeId {
+            fn from(value: &$ty) -> Self {
+                NodeId::$variant(*value)
+            }
+        }
+    };
+}
+
+impl_into_node_id_for!(Expr, ExprId);
+impl_into_node_id_for!(Item, ItemId);
+impl_into_node_id_for!(Stmt, StmtId);
+impl_into_node_id_for!(Body, BodyId);
+impl_into_node_id_for!(Field, FieldId);
+impl_into_node_id_for!(Variant, VariantId);
+
+pub trait Identifiable: Sealed {
```

Comment:
I'm thinking of a convention for such single-method traits. For example, it's custom to name the trait exactly the same as the method that it wraps. Here are examples from std: `From { fn from() }`, `Clone { fn clone() }`, `Default { fn default() }`, `AsRef { fn as_ref() }`. But here are small inconsistencies: `IntoIterator { fn into_iter() }`.

But in this case, if we named the trait just `NodeId` then not only it wouldn't reflect the "verb" of what the trait does, it would also conflict with the `NodeId` enum because they share the same namespace. Maybe as an exception for "getter-like" traits they could be prefixed with `Get` like `trait GetNodeId { fn node_id() }`, and  `Spanned` could be named `trait GetSpan { fn span() }`. 

This is because `Identifiable` and `Spanned` sound rather ad-hoc. For example, if in the future we add a trait for `attributes()` method that would return a list of attributes for the node, then how would we call it `Attributesful` or `Attributed`? Well `Attributted` doesn't sound so bad, but maybe another example for an imaginary `children()` method would it be called `Childrenful`? Naming is hard.. 

Also `Identifiable` sounds like a rather overly generic trait name. What if in the future we add more traits like this, and the new ones will return a new type of ID? With `GetNodeId` such extensibility would be trivial, because the new traits will be called `Get{NewTypeOfIdHere}`.

Reply:
How about `Has<Ty>`, like `HasSpan` and `HasNodeId`? This makes it a property again. I think `Spanned` or `Identifyable` sounds a bit "better" when reading the code, but this won't really work for other things, like the attribute example you mentioned

I believe we took `Spanned` from the example of [`darling::Error::with_span`](https://docs.rs/darling/latest/darling/struct.Error.html#method.with_span).

Reply:
Yeah, I remember rust-analyzer also had a bunch of traits `AttributesOwner` or `CommentsOnwer` etc. Then they renamed them to `HasAttributes` and `HasComments` etc. Idk, either of them works for me just like `GetAttributes`.
Anyway, I'll leave the final judgement call to you